File: python/networking/fetch_weather.py
"""Fetch Current Weather — HTTP client for live weather data from external APIs."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional
import json
import urllib.request
import urllib.error
import urllib.parse


@dataclass
class WindInfo:
    """Wind conditions."""
    speed_mps: float       # meters per second
    direction_deg: float   # compass degrees
    gust_mps: Optional[float] = None

    @property
    def speed_mph(self) -> float:
        return self.speed_mps * 2.23694

    @property
    def speed_kmh(self) -> float:
        return self.speed_mps * 3.6

    @property
    def cardinal(self) -> str:
        directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
                       "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        idx = round(self.direction_deg / 22.5) % 16
        return directions[idx]


@dataclass
class WeatherCondition:
    """A weather condition descriptor."""
    id: int
    main: str          # e.g. "Rain", "Clear", "Clouds"
    description: str   # e.g. "light rain", "clear sky"
    icon: str          # icon code


@dataclass
class WeatherReport:
    """Structured weather report for a location."""
    city: str
    country: str
    latitude: float
    longitude: float
    temperature_k: float    # Kelvin (canonical unit)
    feels_like_k: float
    humidity: int           # percent
    pressure_hpa: int       # hectopascals
    visibility_m: int       # meters
    conditions: list[WeatherCondition]
    wind: WindInfo
    clouds_pct: int         # cloud cover percentage
    timestamp: datetime
    sunrise: Optional[datetime] = None
    sunset: Optional[datetime] = None

    @property
    def temperature_c(self) -> float:
        return self.temperature_k - 273.15

    @property
    def temperature_f(self) -> float:
        return self.temperature_c * 9 / 5 + 32

    @property
    def feels_like_c(self) -> float:
        return self.feels_like_k - 273.15

    @property
    def feels_like_f(self) -> float:
        return self.feels_like_c * 9 / 5 + 32

    @property
    def description(self) -> str:
        if self.conditions:
            return self.conditions[0].description
        return "unknown"

    def summary(self) -> str:
        return (
            f"{self.city}, {self.country}: {self.temperature_c:.1f}°C "
            f"({self.temperature_f:.1f}°F), {self.description}, "
            f"humidity {self.humidity}%, wind {self.wind.speed_mps:.1f} m/s {self.wind.cardinal}"
        )


class WeatherClient:
    """HTTP client for fetching weather data from OpenWeatherMap API.

    Works with any OpenWeatherMap-compatible endpoint. The API key and
    base URL are configurable for testing or alternative providers.
    """

    def __init__(self, api_key: str, base_url: str = "https://api.openweathermap.org") -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")

    def _get(self, path: str, params: dict) -> dict:
        """Make an HTTP GET request and return parsed JSON."""
        params["appid"] = self.api_key
        query = urllib.parse.urlencode(params)
        url = f"{self.base_url}{path}?{query}"
        req = urllib.request.Request(url, headers={"User-Agent": "rosetta-weather/1.0"})
        with urllib.request.urlopen(req, timeout=10) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _parse_report(self, data: dict) -> WeatherReport:
        """Parse OpenWeatherMap JSON into a WeatherReport."""
        sys_data = data.get("sys", {})
        wind_data = data.get("wind", {})
        main_data = data.get("main", {})

        conditions = [
            WeatherCondition(
                id=w["id"],
                main=w["main"],
                description=w["description"],
                icon=w["icon"],
            )
            for w in data.get("weather", [])
        ]

        sunrise = None
        sunset = None
        if "sunrise" in sys_data:
            sunrise = datetime.fromtimestamp(sys_data["sunrise"], tz=timezone.utc)
        if "sunset" in sys_data:
            sunset = datetime.fromtimestamp(sys_data["sunset"], tz=timezone.utc)

        return WeatherReport(
            city=data.get("name", "Unknown"),
            country=sys_data.get("country", "??"),
            latitude=data["coord"]["lat"],
            longitude=data["coord"]["lon"],
            temperature_k=main_data["temp"],
            feels_like_k=main_data["feels_like"],
            humidity=main_data["humidity"],
            pressure_hpa=main_data["pressure"],
            visibility_m=data.get("visibility", 0),
            conditions=conditions,
            wind=WindInfo(
                speed_mps=wind_data.get("speed", 0.0),
                direction_deg=wind_data.get("deg", 0.0),
                gust_mps=wind_data.get("gust"),
            ),
            clouds_pct=data.get("clouds", {}).get("all", 0),
            timestamp=datetime.fromtimestamp(data["dt"], tz=timezone.utc),
            sunrise=sunrise,
            sunset=sunset,
        )

    def by_city(self, city: str, country_code: Optional[str] = None) -> WeatherReport:
        """Fetch current weather by city name."""
        q = f"{city},{country_code}" if country_code else city
        data = self._get("/data/2.5/weather", {"q": q})
        return self._parse_report(data)

    def by_coords(self, lat: float, lon: float) -> WeatherReport:
        """Fetch current weather by latitude and longitude."""
        data = self._get("/data/2.5/weather", {"lat": lat, "lon": lon})
        return self._parse_report(data)

    def by_zip(self, zip_code: str, country_code: str = "US") -> WeatherReport:
        """Fetch current weather by zip/postal code."""
        data = self._get("/data/2.5/weather", {"zip": f"{zip_code},{country_code}"})
        return self._parse_report(data)


def compare_weather(reports: list[WeatherReport]) -> dict:
    """Compare weather across multiple locations.

    Returns a dict with warmest, coldest, most humid, and windiest locations.
    """
    if not reports:
        return {}
    return {
        "warmest": max(reports, key=lambda r: r.temperature_k).city,
        "coldest": min(reports, key=lambda r: r.temperature_k).city,
        "most_humid": max(reports, key=lambda r: r.humidity).city,
        "windiest": max(reports, key=lambda r: r.wind.speed_mps).city,
    }


# --- demo ---
if __name__ == "__main__":
    import os

    key = os.environ.get("OWM_API_KEY", "")
    if not key:
        print("Set OWM_API_KEY environment variable to run the demo.")
        print("Example: OWM_API_KEY=your_key python fetch_weather.py")
    else:
        client = WeatherClient(key)
        report = client.by_city("Jacksonville", "US")
        print(report.summary())

File: go/networking/fetch_weather.go
// Fetch Current Weather — HTTP client for live weather data from external APIs.
package networking

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"time"
)

// WindInfo holds wind conditions.
type WindInfo struct {
	SpeedMps     float64  `json:"speed"`
	DirectionDeg float64  `json:"deg"`
	GustMps      *float64 `json:"gust,omitempty"`
}

// SpeedMph returns wind speed in miles per hour.
func (w WindInfo) SpeedMph() float64 { return w.SpeedMps * 2.23694 }

// SpeedKmh returns wind speed in km/h.
func (w WindInfo) SpeedKmh() float64 { return w.SpeedMps * 3.6 }

// Cardinal returns the compass direction as a string.
func (w WindInfo) Cardinal() string {
	dirs := [16]string{
		"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
		"S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW",
	}
	idx := int(math.Round(w.DirectionDeg/22.5)) % 16
	return dirs[idx]
}

// WeatherCondition describes a weather state.
type WeatherCondition struct {
	ID          int    `json:"id"`
	Main        string `json:"main"`
	Description string `json:"description"`
	Icon        string `json:"icon"`
}

// WeatherReport holds structured weather data for a location.
type WeatherReport struct {
	City         string
	Country      string
	Latitude     float64
	Longitude    float64
	TemperatureK float64
	FeelsLikeK   float64
	Humidity     int
	PressureHpa  int
	VisibilityM  int
	Conditions   []WeatherCondition
	Wind         WindInfo
	CloudsPct    int
	Timestamp    time.Time
	Sunrise      *time.Time
	Sunset       *time.Time
}

// TemperatureC returns temperature in Celsius.
func (r WeatherReport) TemperatureC() float64 { return r.TemperatureK - 273.15 }

// TemperatureF returns temperature in Fahrenheit.
func (r WeatherReport) TemperatureF() float64 { return r.TemperatureC()*9.0/5.0 + 32.0 }

// FeelsLikeC returns feels-like in Celsius.
func (r WeatherReport) FeelsLikeC() float64 { return r.FeelsLikeK - 273.15 }

// FeelsLikeF returns feels-like in Fahrenheit.
func (r WeatherReport) FeelsLikeF() float64 { return r.FeelsLikeC()*9.0/5.0 + 32.0 }

// Description returns the primary weather description.
func (r WeatherReport) Description() string {
	if len(r.Conditions) > 0 {
		return r.Conditions[0].Description
	}
	return "unknown"
}

// Summary returns a one-line summary.
func (r WeatherReport) Summary() string {
	return fmt.Sprintf(
		"%s, %s: %.1f°C (%.1f°F), %s, humidity %d%%, wind %.1f m/s %s",
		r.City, r.Country, r.TemperatureC(), r.TemperatureF(),
		r.Description(), r.Humidity, r.Wind.SpeedMps, r.Wind.Cardinal(),
	)
}

// owmResponse mirrors the OpenWeatherMap JSON response.
type owmResponse struct {
	Coord struct {
		Lat float64 `json:"lat"`
		Lon float64 `json:"lon"`
	} `json:"coord"`
	Weather []WeatherCondition `json:"weather"`
	Main    struct {
		Temp      float64 `json:"temp"`
		FeelsLike float64 `json:"feels_like"`
		Pressure  int     `json:"pressure"`
		Humidity  int     `json:"humidity"`
	} `json:"main"`
	Visibility int      `json:"visibility"`
	Wind       WindInfo `json:"wind"`
	Clouds     struct {
		All int `json:"all"`
	} `json:"clouds"`
	Dt  int64 `json:"dt"`
	Sys struct {
		Country string `json:"country"`
		Sunrise int64  `json:"sunrise"`
		Sunset  int64  `json:"sunset"`
	} `json:"sys"`
	Name string `json:"name"`
}

// WeatherClient fetches weather data from an OpenWeatherMap-compatible API.
type WeatherClient struct {
	APIKey  string
	BaseURL string
	client  *http.Client
}

// NewWeatherClient creates a client with the given API key.
func NewWeatherClient(apiKey string) *WeatherClient {
	return &WeatherClient{
		APIKey:  apiKey,
		BaseURL: "https://api.openweathermap.org",
		client:  &http.Client{Timeout: 10 * time.Second},
	}
}

func (c *WeatherClient) fetch(params url.Values) (*WeatherReport, error) {
	params.Set("appid", c.APIKey)
	u := fmt.Sprintf("%s/data/2.5/weather?%s", c.BaseURL, params.Encode())

	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}
	req.Header.Set("User-Agent", "rosetta-weather/1.0")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetch: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read body: %w", err)
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("API error %d: %s", resp.StatusCode, string(body))
	}

	var raw owmResponse
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, fmt.Errorf("parse JSON: %w", err)
	}
	return parseReport(&raw), nil
}

func parseReport(raw *owmResponse) *WeatherReport {
	ts := time.Unix(raw.Dt, 0).UTC()
	var sunrise, sunset *time.Time
	if raw.Sys.Sunrise > 0 {
		t := time.Unix(raw.Sys.Sunrise, 0).UTC()
		sunrise = &t
	}
	if raw.Sys.Sunset > 0 {
		t := time.Unix(raw.Sys.Sunset, 0).UTC()
		sunset = &t
	}
	return &WeatherReport{
		City:         raw.Name,
		Country:      raw.Sys.Country,
		Latitude:     raw.Coord.Lat,
		Longitude:    raw.Coord.Lon,
		TemperatureK: raw.Main.Temp,
		FeelsLikeK:   raw.Main.FeelsLike,
		Humidity:     raw.Main.Humidity,
		PressureHpa:  raw.Main.Pressure,
		VisibilityM:  raw.Visibility,
		Conditions:   raw.Weather,
		Wind:         raw.Wind,
		CloudsPct:    raw.Clouds.All,
		Timestamp:    ts,
		Sunrise:      sunrise,
		Sunset:       sunset,
	}
}

// ByCity fetches weather for a city name, with optional country code.
func (c *WeatherClient) ByCity(city string, countryCode string) (*WeatherReport, error) {
	q := city
	if countryCode != "" {
		q = city + "," + countryCode
	}
	params := url.Values{"q": {q}}
	return c.fetch(params)
}

// ByCoords fetches weather for a latitude/longitude pair.
func (c *WeatherClient) ByCoords(lat, lon float64) (*WeatherReport, error) {
	params := url.Values{
		"lat": {fmt.Sprintf("%f", lat)},
		"lon": {fmt.Sprintf("%f", lon)},
	}
	return c.fetch(params)
}

// ByZip fetches weather for a zip/postal code.
func (c *WeatherClient) ByZip(zip, countryCode string) (*WeatherReport, error) {
	params := url.Values{"zip": {zip + "," + countryCode}}
	return c.fetch(params)
}

// CompareWeather returns the warmest, coldest, most humid, and windiest
// locations from a set of reports.
func CompareWeather(reports []WeatherReport) map[string]string {
	result := make(map[string]string)
	if len(reports) == 0 {
		return result
	}
	warmest, coldest, humid, windy := &reports[0], &reports[0], &reports[0], &reports[0]
	for i := range reports {
		r := &reports[i]
		if r.TemperatureK > warmest.TemperatureK {
			warmest = r
		}
		if r.TemperatureK < coldest.TemperatureK {
			coldest = r
		}
		if r.Humidity > humid.Humidity {
			humid = r
		}
		if r.Wind.SpeedMps > windy.Wind.SpeedMps {
			windy = r
		}
	}
	result["warmest"] = warmest.City
	result["coldest"] = coldest.City
	result["most_humid"] = humid.City
	result["windiest"] = windy.City
	return result
}
